### PR TITLE
Remove unnecessary sorting for highestSupportedVersion

### PR DESCRIPTION
**Pull request by tedyu**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
The sorting step in highestSupportedVersion is not needed.
The searching and error handling can be expressed in linear complexity.

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: pkg/volume/csi/csi_plugin.go [hunk 1]
```diff
@@ -892,34 +891,33 @@ func highestSupportedVersion(versions []string) (*utilversion.Version, error) {
...
+	var theErr error
 	for i := len(versions) - 1; i >= 0; i-- {
-		highestSupportedVersion, err := utilversion.ParseGeneric(versions[i])
+		currentHighestVer, err := utilversion.ParseGeneric(versions[i])
 		if err != nil {
-			return nil, err
+			theErr = err
+			continue
+		} else {
+			seenNoErr = true
```
Comment (saad-ali) [inline pkg/volume/csi/csi_plugin.go]: Why is this necessary?

File: pkg/volume/csi/csi_plugin.go [hunk 2]
```diff
@@ -892,34 +891,33 @@ func highestSupportedVersion(versions []string) (*utilversion.Version, error) {
...
+			highestSupportedVersion = currentHighestVer
+		} else if highestSupportedVersion.LessThan(currentHighestVer) {
+			highestSupportedVersion = currentHighestVer
 		}
 	}
 
+	if highestSupportedVersion != nil {
+		return highestSupportedVersion, nil
+	}
+	if theErr != nil && !seenNoErr {
```
Comment (saad-ali) [inline pkg/volume/csi/csi_plugin.go]: Previously this code didn't return the parsing error. Do we really want to change that? Can we instead expand the generic error below to include 1) the list of versions returned by the driver and 2) any errors in parsing?

File: pkg/volume/csi/csi_plugin.go (hunk 1 above)
Comment (tedyu) [reply pkg/volume/csi/csi_plugin.go]: seenNoErr is used to record whether there is a parsable version in the input.

**Comment by tedyu** [comment]
@saad-ali 
seenNoErr has been removed.
Please take another look.

File: pkg/volume/csi/csi_plugin.go [hunk 3]
```diff
@@ -892,35 +891,28 @@ func highestSupportedVersion(versions []string) (*utilversion.Version, error) {
...
+		} else if highestSupportedVersion.LessThan(currentHighestVer) {
+			highestSupportedVersion = currentHighestVer
 		}
 	}
 
-	return nil, errors.New(log("None of the CSI versions reported by this driver are supported"))
+	if highestSupportedVersion != nil {
+		return highestSupportedVersion, nil
+	}
+	return nil, errors.New(fmt.Sprintf("None of the CSI versions reported by this driver are supported : %v", theErr))
```
Comment (saad-ali) [inline pkg/volume/csi/csi_plugin.go]: nit: can you also add emit `versions` so that it is clear what the driver returned that Kubernetes didn't like..

**Comment by tedyu** [comment]
@saad-ali 
Please take another look.

**Review by saad-ali — APPROVED** [review]
/approve

Will let @taragu LGTM

**Comment by saad-ali** [comment]
```
Errors from golint:
pkg/volume/csi/csi_plugin.go:915:14: should replace errors.New(fmt.Sprintf(...)) with fmt.Errorf(...)
```

**Comment by tedyu** [comment]
@saad-ali 
Thanks for noticing golint warning.

It has been addressed.

File: pkg/volume/csi/csi_plugin.go [hunk 4]
```diff
@@ -892,35 +891,28 @@ func highestSupportedVersion(versions []string) (*utilversion.Version, error) {
...
+		currentHighestVer, err := utilversion.ParseGeneric(versions[i])
 		if err != nil {
-			return nil, err
+			theErr = err
+			continue
 		}
-
-		if highestSupportedVersion.Major() <= 1 {
-			return highestSupportedVersion, nil
+		if currentHighestVer.Major() > 1 {
```
Comment (taragu) [inline pkg/volume/csi/csi_plugin.go]: @saad-ali do you know why we are skipping major version > 1? Do we want to make it a const?

File: pkg/volume/csi/csi_plugin.go [hunk 5]
```diff
@@ -892,35 +891,28 @@ func highestSupportedVersion(versions []string) (*utilversion.Version, error) {
...
+			continue
+		}
+		if highestSupportedVersion == nil {
```
Comment (taragu) [inline pkg/volume/csi/csi_plugin.go]: Nit: `if highestSupportedVersion == nil || highestSupportedVersion.LessThan(currentHighestVer){`

File: pkg/volume/csi/csi_plugin.go (hunk 4 above)
Comment (tedyu) [reply pkg/volume/csi/csi_plugin.go]: If we define a constant, would it be defined in semver.go ?

@saad-ali 
What do you think ?

Comment (saad-ali) [reply pkg/volume/csi/csi_plugin.go]: I believe the logic here was that CSI currently only has version 0.x and 1.x (see https://github.com/container-storage-interface/spec/releases). And therefore any driver claiming version 2.x+ is ignored as an unsupported versions. This is ok because future 1.x versions of CSI are supposed to be backwards compatible so this version of Kubernetes will work with any 1.x driver (or 0.x), but it may not work with 2.x drivers (because 2.x does not have to be backwards compatible with 1.x).

Making this a const, SGTM. Also, maybe add a comment capturing this, because I'm sure others reading the code will have the same question in the future.

**Comment by tedyu** [comment]
@taragu 
Please take another look.
